### PR TITLE
ARROW-2822: [C++] Zero padding bytes in PoolBuffer

### PR DESCRIPTION
**Pull request by alendit**
See the [JIRA discussion](https://issues.apache.org/jira/projects/ARROW/issues/ARROW-2822).

### CODE REVIEW

File: cpp/src/arrow/builder.cc
Comment (pitrou) [inline cpp/src/arrow/builder.cc]: Is this desirable? If an error occurs while building an array, it is normal not to finish the builder.

File: cpp/src/arrow/builder.cc
Comment (pitrou) [inline cpp/src/arrow/builder.cc]: Note null buffers are problematic, see PR #2243.

File: cpp/src/arrow/buffer.h
Comment (pitrou) [inline cpp/src/arrow/buffer.h]: Not a native English speaker, but "between padding" looks grammatically incorrect.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for doing this! On the principle this looks sane. A couple comments below.

File: cpp/src/arrow/buffer.h
Comment (wesm) [reply cpp/src/arrow/buffer.h]: "in" instead of "between" sounds fine

File: cpp/src/arrow/builder.cc
Comment (wesm) [reply cpp/src/arrow/builder.cc]: How about hiding this behind a flag like `ARROW_LOG_UNFINISHED_BUILDERS`. Having this output in debug builds always seems like too much to me

File: cpp/src/arrow/builder.cc
Comment (wesm) [inline cpp/src/arrow/builder.cc]: What's the rationale for this change? All values in NullBuilder are null

File: cpp/src/arrow/builder.cc
Comment (wesm) [inline cpp/src/arrow/builder.cc]: Ditto re: decision above

File: cpp/src/arrow/builder.cc
Comment (wesm) [inline cpp/src/arrow/builder.cc]: Is setting unfinished here needed?

File: cpp/src/arrow/builder.cc
Comment (wesm) [inline cpp/src/arrow/builder.cc]: Curious why this is needed

File: cpp/src/arrow/builder.h
Comment (wesm) [inline cpp/src/arrow/builder.h]: Seems like this should only be set when the builder is initialized

File: cpp/src/arrow/io/file.cc
Comment (wesm) [inline cpp/src/arrow/io/file.cc]: It seems error prone to have to remember to zero the padding. What would be the performance consequences (if any, realistically) of zeroing in Resize?

File: cpp/src/arrow/python/python_to_arrow.cc
Comment (wesm) [inline cpp/src/arrow/python/python_to_arrow.cc]: the way that `NullBuilder` is being used in this module doesn't look right; probably related to the NullBuilder changes above cc @pcmoritz. It looks like a BooleanBuilder should be used instead

File: cpp/src/arrow/io/file.cc
Comment (pitrou) [reply cpp/src/arrow/io/file.cc]: Some things may be repeatedly resized if you don't know the exact size up front (e.g. in BinaryBuilder).

Comment (wesm) [reply cpp/src/arrow/io/file.cc]: Right, in the event that you shrank the buffer in `BinaryBuilder` when `Finish` is called, you would want to zero the padding there, too

Comment (wesm) [reply cpp/src/arrow/io/file.cc]: I guess if you shrink the memory then the memory you padded before goes unused

Comment (alendit) [reply cpp/src/arrow/io/file.cc]: The worst case is when you Reserve a huge amount of memory and then gradually Resize it, as the new elements come in. You would expect these operations to be virtually free while in reality you are zeroing everything between size and capacity on every call.

Comment (wesm) [reply cpp/src/arrow/io/file.cc]: Got it. OK, let's leave things as they are, then

File: cpp/src/arrow/builder.cc
Comment (alendit) [reply cpp/src/arrow/builder.cc]: You want to rework it and move the warning to `null_bitmap` and `data` methods. It seems to be doable, but I'd need some more time. 

We can still then move it behind a special compile flag.

File: cpp/src/arrow/buffer.h
Comment (alendit) [reply cpp/src/arrow/buffer.h]: Yeah, comment editing fail, I'll fix it.

File: cpp/src/arrow/builder.cc
Comment (alendit) [reply cpp/src/arrow/builder.cc]: Oh, then I've misunderstood it's purpose. I thought it should be used as in `python_to_arrow.cc`. 

I'll revert it and add some unit cases, you could look at.

File: cpp/src/arrow/builder.cc
Comment (alendit) [reply cpp/src/arrow/builder.cc]: Since unfinished signals that there were values appended after the last `FinishInternal` call, I would say yes.

File: cpp/src/arrow/builder.cc
Comment (alendit) [reply cpp/src/arrow/builder.cc]: There are nulls appended to the `values_builder_`, so it could resized and contain uninitialized padding, if we never call finish on it.

File: cpp/src/arrow/builder.h
Comment (alendit) [reply cpp/src/arrow/builder.h]: This was my first try. Then I hit the problem that sometimes the builders are initialized but never appended to and never finished, i.e. in DictionaryBuilder. So I changed the logic to "set to false on every append". 

I'll see if I can change it back after the rework I've mentioned before.

**Comment by alendit** [comment]
Hi guys. I'll try to incorporate the changes and rework the warning logic to be more precise in the comming days (hopefully tomorrow).

**Comment by alendit** [comment]
Ok, so there was a possibility I totally haven't considered before: we can just remove the `data()` and `null_bitmap` from the builders. Amazingly, it doesn't break much besides couple of tests and the `DictionaryBuilder`. I adjusted `DictionaryBuilder` and it seems to work like a charm.

Does anyone know, if some 3rd party code relies on these methods? I would argue, they should be considered implementation detail, but we might want to put a deprecation warning before removing them alltogether.

File: cpp/src/arrow/python/python_to_arrow.cc
Comment (alendit) [reply cpp/src/arrow/python/python_to_arrow.cc]: I reverted the `NullBuilder` changes and switched `nones_` to `BooleanBuilder`.

@pcmoritz could you check if 6753dcc is fine?

**Comment by pitrou** [comment]
> Does anyone know, if some 3rd party code relies on these methods? I would argue, they should be considered implementation detail, but we might want to put a deprecation warning before removing them alltogether.

I would deprecate them first indeed.

**Comment by alendit** [comment]
Rebased on the master.

File: cpp/src/arrow/adapters/orc/adapter.cc
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter.cc]: Hmm... I think it's a bit problematic that we seem to be introducing a second data copy here (first from `source` to `converted`, then from `converted` to the builder). Can we avoid the second copy without using `builder->data()`? Otherwise we may have to undeprecate the latter.

File: cpp/src/arrow/buffer.h
Comment (pitrou) [inline cpp/src/arrow/buffer.h]: I think "fixed-size" is the correct spelling. Perhaps a native English speaker can confirm?

File: cpp/src/arrow/buffer.h
Comment (pitrou) [inline cpp/src/arrow/buffer.h]: I think the article here ("a resizeable buffer") shouldn't be removed.

**Review by pitrou — CHANGES_REQUESTED** [review]
Looks like the deprecation may be problematic after all. Sorry for all the back-and-forth!

File: cpp/src/arrow/adapters/orc/adapter.cc
Comment (wesm) [inline cpp/src/arrow/adapters/orc/adapter.cc]: stack allocation here

File: cpp/src/arrow/buffer.h
Comment (wesm) [reply cpp/src/arrow/buffer.h]: I googled a bit.  "fixed size" seems more common, and reads OK to me

File: cpp/src/arrow/builder.cc
Comment (wesm) [reply cpp/src/arrow/builder.cc]: I just left a review + comments there. I think we should try to merge this before merging #2243 so we can decide what to do there

File: cpp/src/arrow/builder.h
Comment (wesm) [inline cpp/src/arrow/builder.h]: Are these lines equivalent (I thought so)?

File: cpp/src/arrow/builder.h
Comment (wesm) [inline cpp/src/arrow/builder.h]: typo

File: cpp/src/arrow/builder.h
Comment (wesm) [inline cpp/src/arrow/builder.h]: typo

File: cpp/src/arrow/builder.h
Comment (wesm) [inline cpp/src/arrow/builder.h]: typo

File: cpp/src/arrow/builder.h
Comment (wesm) [inline cpp/src/arrow/builder.h]: If this is trivial, it can be omitted altogether?

File: cpp/src/arrow/python/python_to_arrow.cc
Comment (wesm) [inline cpp/src/arrow/python/python_to_arrow.cc]: @pcmoritz or @robertnishihara take a look to make sure these changes look right. 

Should this be `Append(false)`? I would expect `nones_` to be an all non-null boolean array

File: python/pyarrow/tests/test_builder.py
Comment (wesm) [inline python/pyarrow/tests/test_builder.py]: Is this still needed?

**Review by wesm — COMMENTED** [review]
It's good to make all this more precise + clean things up. I left some comments / questions

File: cpp/src/arrow/adapters/orc/adapter.cc
Comment (wesm) [reply cpp/src/arrow/adapters/orc/adapter.cc]: Note we don't have any unit tests for this code path so have to be extra careful

**Comment by alendit** [comment]
So, what orc adapter did was to use `std::copy` to copy data from `int64_t*` input to a `Int32` builder by specifying `data()` as a target. This works because `std::copy` takes iterators and they just need to conform to the interator interface. IThe builders, on the other hand, either accept a vector or a raw pointer, both have to have the exact expected type to work.

My solution is to overload the `AppendValues` method to accept iterators. Then we can just use `std::copy` internally. The number of existing overloads with all possible combination of pointer and vector arguments suggests that we might want to switch to iterator-accepting interfaces at some point in the future, since they are more general. In the ideal world we'd have ranges, but alas...

An additional problem was that orc adapter used data pointer to directly add a computed value (time in nanoseconds). What is easily done in python with generators, is not supported in any current C++ version directly (excluding non-standard extensions). I added the `LazyRange`, a thin wrapper around a callable (e.g. lambda) which returns objects conforming to the iterator interface. What looks like a complicated construct, can be optimized by the compiler very well (I added some benchmarks). 

I really think we should remove/deprecate `data()` and `null_bitmap()` from the builders, but it's apparent, we'd need some interface restructuring. I think the changes would be beneficial in the long term, but I'm also not happy about the size of this diff. I would suggest that as long as the general direction is acceptable, we fix this PR's code and leave the more farreaching consideration to be fixed in further PRs.

**Comment by alendit** [comment]
Ok, so funny thing: MSVC14 (i.e. 2015) issues warning on narrowing conversion when using `std::copy`. The funny part is that none other compiler, not even later MSVC versions, seem to be bothered by it. 

Which raises the question, if the orc adapter in its current form has even been through CI, because [here|https://github.com/apache/arrow/blob/master/cpp/src/arrow/adapters/orc/adapter.cc#L535] it totaly copies `int64_t` to `int32_t` and smaller. Of course, it's not a problem, since we check the supplied type beforehand, but it shouldn't compile with MSVC 2015.

With the supplied changes it's actually pretty easy to fix the warning: just wrap the iterator with a lambda which does a `static_cast` to the right type. Should I do it? Should be 0 costs when compiled on `Release`.

**Comment by pitrou** [comment]
@alendit can you confirm when the PR is ready for review again? Thanks for working on this!

**Comment by wesm** [comment]
>  I would suggest that as long as the general direction is acceptable, we fix this PR's code and leave the more farreaching consideration to be fixed in further PRs.

Sounds good to me

**Comment by wesm** [comment]
Reviewing this again now

File: cpp/src/arrow/adapters/orc/adapter.cc
Comment (wesm) [inline cpp/src/arrow/adapters/orc/adapter.cc]: Cool

File: cpp/src/arrow/builder.h
Comment (wesm) [inline cpp/src/arrow/builder.h]: As an aside, would you expect the optimized code to be significantly different here vs. the original `const uint8_t* valid_bytes` version?

**Review by wesm — APPROVED** [review]
+1, wow thanks @alendit for going to the effort of adding iterator versions for all the AppendValues functions, definitely going above and beyond. This will be really helpful going forward.

I took a pretty careful read through this and it looks good to me. @pitrou do you want to look any more or do you mind if I go ahead and merge? https://github.com/apache/arrow/pull/2243 will need to be rebased after but it shouldn't be too bad

File: cpp/src/arrow/util/lazy-benchmark.cc
Comment (pitrou) [inline cpp/src/arrow/util/lazy-benchmark.cc]: I am curious: what do you call a lazy iterator? ISTM iterators are always lazy.

File: cpp/src/arrow/builder.h
Comment (wesm) [reply cpp/src/arrow/builder.h]: Since these are always C types I would think that just `value_type` is fine; I am not an expert on the fine details, though

File: cpp/src/arrow/util/lazy-benchmark.cc
Comment (wesm) [reply cpp/src/arrow/util/lazy-benchmark.cc]: I believe he's referring to the object returned by `internal::MakeLazyRange`

**Review by pitrou — APPROVED** [review]

**Comment by wesm** [comment]
:+1: ok, I'm merging so we can get your other PR rebased. Thanks @alendit @pitrou for your careful attention on this issue

File: cpp/src/arrow/builder.h
Comment (alendit) [reply cpp/src/arrow/builder.h]: With STL iterators and raw pointers it should be exactly the same. With more complicated iterators, it's hard to tell, though the `LazyRange` ones the same performance as loops, at least on GCC.

File: cpp/src/arrow/builder.h
Comment (alendit) [reply cpp/src/arrow/builder.h]: Yupp, the first const should go.

File: cpp/src/arrow/util/lazy-benchmark.cc
Comment (alendit) [reply cpp/src/arrow/util/lazy-benchmark.cc]: Yes, the `LazyRange` is lazy about calling the supplied generator function instead of generating a new container eagerly like `std::generate`, for example. 

With these benchmarks I wanted to make sure, that the compilers can inline the lambdas properly so we don't sacrifice performance compared to a loop.

**Comment by alendit** [comment]
Nice, glad it went through. Thx both of you for the help and patience :) 

Once further question: for really small changes, like removing that one `const`, do I open a JIRA issue, too? Or is there a simplified procedure, so to not to swamp the queue?

**Comment by wesm** [comment]
You can open a PR but don't bother creating a JIRA. I'll merge once the build runs

**Pull request closed without merging**
